Add an icon and visually hidden a11y text to links that open in a new tab

talldan opened the pull request:
## Description
Closes #1105 

Adds an icon and the visually hidden text '(Opens in a new window)' for improved a11y to two links that open in a new tab.

- Use `ExternalLink` component for the link editor toolbar option on the Paragraph block. It replaces an `a` tag. This adds an additional icon and the visibly hidden text to the url.
- Use the `ExternalLink` component for the permalink url when editing the post header

Furthermore a change has been made to the preview button in the top toolbar so that is a button element instead of an anchor.

## How has this been tested?
- Manually tested visual appearance of icon
- Manually tested 'Opens in a new window' a11y text using VoiceOver
- Ensured automated tests pass

## Screenshots <!-- if applicable -->
Permalink:
![screen shot 2018-07-17 at 11 14 39 am](https://user-images.githubusercontent.com/677833/42794675-bf91b424-89b2-11e8-8130-e9def12f2c81.png)


Preview (visually unchanged, but is now a button element):
![screen shot 2018-07-17 at 11 14 56 am](https://user-images.githubusercontent.com/677833/42794687-cd771660-89b2-11e8-82c5-605b6176c24d.png)


Paragraph Block Link:
![screen shot 2018-07-17 at 11 14 50 am](https://user-images.githubusercontent.com/677833/42794696-dd7784d2-89b2-11e8-871a-c72e736e04ea.png)


## Types of changes
<!-- What types of changes does your code introduce?  -->
Bug fix (non-breaking change which fixes an issue)
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

talldan committed:
Use ExternalLink for link in editor toolbar for RichText block

Links added in RichText are opened in a new tab when using the editor toolbar.
This commit switches the anchor tag with an ExternalLink component. This makes
the link more accessible (it adds some hidden text) and provides an icon
indicating that the link is external.

talldan committed:
Add prop for icon class name to ExternalLink

talldan committed:
Allow Button component to render an ExternalLink

talldan committed:
Specify PostPreviewButton as opening external links

talldan committed:
Fix overly specific selector preventing override of icon styles

talldan committed:
Position dashicon for external links when rendered via Button

talldan committed:
Set isExternalLink true for post permalink and adjust styling

talldan committed:
Fix incorrect default display property given to button

talldan committed:
Use flex to center icon in button and add a tiny bit of margin to adjust

talldan commented:
@karmatosed Would be great if you could just double-check the screenshots look ok. The only thing I spotted is that the link text color isn't the same for the permalink and the rich text link. I can make a separate issue if you think that's worth making consistent.

talldan commented on components/button/style.scss:
```diff
@@ -181,6 +181,15 @@
 		font-size: 11px;
 	}
+
+	.components-external-link__icon {
```
I wasn't entirely sure that these are the right selectors to use, or whether I should somehow specify a class of components-button__icon on the icon.

noisysocks replied:
The class should begin with `components-button__` since the node is in the Button component.

talldan commented:
ah - I need to add some tests, particularly on the button element. Will add some asap.

noisysocks commented:
Very supportive of the changes to the link formatting tool and permalink editor here. It looks great 👍 

I'm not so sure about changing the Preview button, though.

In my mind<sup>1</sup>, the point of the external icon is to shape the user's expectation so that they're not surprised when a new tab opens. This makes total sense since the expectation when clicking a hyperlink is that the browser navigates to a new page in-place. 

But I don't think users have any such expectations when they click on a button. A button doesn't imply that clicking it will navigate to a new page. It's a button—buttons can do anything!

Could an alternative to this be to make the Preview button a `<button>` instead of an `<a>`? Or, could we add we add `aria-role="button"` and have it open when a user selects it and presses Space?

<sup>1</sup> Read: I am not a UX designer! 🙂

talldan committed:
Add tests for isExternalLink prop in Button component

talldan committed:
Pass component-button__icon class through to icon, and use it for scss selectors

karmatosed reviewed (changes requested):
@noisysocks is hereby granted temporary 'UX designer' status as he was on point with suggesting this doesn't get done to the preview button :) I would say apart from that so long as things are just links (not buttons hiding in link clothing) this is a great improvement.

talldan committed:
Make post preview button render as a button instead of a link

For accessibility purposes the post preview button is now rendered as a button,
meaning it can be interacted with using the keyboard (space bar to open a
preview page).

afercia commented:
> and I'm wary of Button becoming a component that has too many responsibilities.

This is the main reason why a previous attempt in #6786 was abandoned. It ended up in adding a lot of stuff to the Button component. Currently, the Button component is named based on its visual appearance. This is far from ideal and introduces some potential confusion. Components should be named based on what they do. In the discussion on https://github.com/WordPress/gutenberg/pull/6786#issuecomment-392903261 at some point it was agreed to split the Button component in two separate components, in short:
- Link: it's an `<a>` element, used to navigate to a web resource (i.e. no `href="#"` links used to "do something") and visually styled as a link or a button
- Button: it's a `<button>` element, used to "do something" and visually styled as a button or link

Based on that #7534 was created to separate the two components _and eliminate ExternalLink_. I know that takes a different path from the work done here, sorry for that, but I'd strongly encourage to focus on the effort proposed on #7534.

talldan committed:
Revert "Pass component-button__icon class through to icon, and use it for scss selectors"

This reverts commit 77422fce000a2b75788b5fadaa1cd26e770799e3.

Revert "Add tests for isExternalLink prop in Button component"

This reverts commit 5102826e2c22c7ae59c8d21a3b1fb44781babe5f.

Revert "Use flex to center icon in button and add a tiny bit of margin to adjust"

This reverts commit 0f3c5fa792dc3386e05acd3821ea32b245b42377.

Revert "Set isExternalLink true for post permalink and adjust styling"

This reverts commit 402ecd3ff93f3b945072124f4bd47e15fb3470ce.

Revert "Position dashicon for external links when rendered via Button"

This reverts commit 0930efac6c9eb709b5af7e0e28e9d05d3346bebd.

Revert "Allow Button component to render an ExternalLink"

This reverts commit 1b8e9ca5c5fe67aab5d08d2ffc0798fb7edfaa8d.

Revert "Add prop for icon class name to ExternalLink"

This reverts commit ddac73dd0f09a1076f842fdb89ece370e274ac82.

talldan committed:
Make the post permalink link an ExternalLink instead of a Button

talldan committed:
Rename member variable used as ref to reflect that it's not longer a button

talldan committed:
Modify ExternalLink component to forwards its ref

talldan committed:
Adjust styling to work in a wider range of contexts

noisysocks reviewed (approved):
Aside from the failing test this looks great! 🙂

noisysocks commented:
It looks like the failing test was just an intermittent E2E test failure. I restarted the job for you.

talldan commented:
Great - thanks @noisysocks.

@karmatosed - I've updated the screenshots at the top, are you happy to approve this now?

karmatosed commented:
@talldan do we need to have the link icon in the actual link interface? That feels over the top to me.

talldan commented:
@karmatosed - do you mean this icon? ![external link](https://user-images.githubusercontent.com/253067/42348396-351cf246-80a1-11e8-977c-87ff43f9f9c5.png)

I thought that was one of the main points we were trying to address from #1105. My opinion is that we should use the icon consistently on anything that looks like and interacts like a link.

karmatosed commented:
@talldan my understanding was that it was for interface links not links in the drop down. It's this bit I am questioning:

![2018-07-18 at 14 12](https://user-images.githubusercontent.com/253067/42883840-a5e0c8ca-8a94-11e8-8025-b926eb42de60.png)

If everyone wants that's ok it just feels a little odd seeing beside something like this.

karmatosed commented:
Let's add it in @talldan and iterate.
